cdap-ranger/cdap-ranger-binding/src/main/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizer.java

```diff
@@ -215,6 +179,61 @@ public void removeRoleFromPrincipal(Role role, Principal principal) throws Excep
     return null;
   }
 
+  private boolean enforce(EntityId entity, Principal principal,
+                          RangerAccessRequest.ResourceMatchingScope resourceMatchingScope, String accessType)
+    throws Exception {
+    LOG.debug("Enforce called on entity {}, principal {}, action {} and match scope {}", entity, principal,
+              accessType, resourceMatchingScope);
+    if (rangerPlugin == null) {
+      throw new RuntimeException("CDAP Ranger Authorizer is not initialized.");
+    }
+
+    if (principal.getType() != Principal.PrincipalType.USER) {
+      throw new IllegalArgumentException(String.format("The principal type for current enforcement request is '%s'. " +
+                                                         "Authorization enforcement is only supported for '%s'.",
+                                                       principal.getType(), Principal.PrincipalType.USER));
+    }
+    String requestingUser = principal.getName();
+    String ip = InetAddress.getLocalHost().getHostName();
+    Set<String> userGroups = MiscUtil.getGroupsForRequestUser(requestingUser);
```

Comment:
When doing multiple enforce requests as part of isVisible, we can compute all these including date only once per request. Since you already mentioned that you want to use batch requests later, we can make this change at that time.